### PR TITLE
Add Algorithmic Complexity Info

### PR DESCRIPTION
**Pull request by MrBurmark**
# Add Algorithmic Complexity Info

Add the ability to print the algorithmic complexity of each kernel.

- This PR is a feature
- It does the following:
  - Adds algorithmic complexity info at the request of me and @pearce8

### CODE REVIEW

**Comment by MrBurmark** [comment]
@michaelmckinsey1 Can you take a look at the caliper stuff, I put a draft in but was completely guessing as to how to add strings.

File: src/common/KernelBase.cpp [hunk 1]
```diff
@@ -99,6 +101,10 @@ KernelBase::KernelBase(KernelID kid, const RunParams& params)
                                               CALI_ATTR_AGGREGATABLE |
                                               CALI_ATTR_SKIP_EVENTS);
   }
+  Complexity_attr = cali_create_attribute("Complexity", CALI_TYPE_STRING,
+                                           CALI_ATTR_ASVALUE |
+                                           CALI_ATTR_AGGREGATABLE |
```
Comment (michaelmckinsey1) [inline src/common/KernelBase.cpp:106]:
```suggestion
```

File: src/common/KernelBase.cpp [hunk 2]
```diff
@@ -574,6 +581,7 @@ void KernelBase::doOnceCaliMetaBegin(VariantID vid, size_t tune_idx)
         std::string feature = getFeatureName(fid);
         cali_set_int(Feature_attrs[feature], usesFeature(fid));
     }
+    cali_set_string(Complexity_attr, getComplexityName(getComplexity()));
```
Comment (michaelmckinsey1) [inline src/common/KernelBase.cpp:584]:
```suggestion
    cali_set_string(Complexity_attr, getComplexityName(getComplexity()).c_str());
```

File: src/common/KernelBase.cpp [hunk 3]
```diff
@@ -624,7 +632,8 @@ void KernelBase::setCaliperMgrVariantTuning(VariantID vid,
           { "expr": "any(max#Reduction)", "as": "FeatureReduction" },
           { "expr": "any(max#Atomic)", "as": "FeatureAtomic" },
           { "expr": "any(max#View)", "as": "FeatureView" },
-          { "expr": "any(max#MPI)", "as": "FeatureMPI" }
+          { "expr": "any(max#MPI)", "as": "FeatureMPI" },
+          { "expr": "any(max#Complexity)", "as": "Complexity" }
         ]
```
Comment (michaelmckinsey1) [inline src/common/KernelBase.cpp:637]:
```suggestion
        ],
        "group by": ["Complexity"],
```

File: src/common/KernelBase.cpp [hunk 4]
```diff
@@ -650,7 +659,8 @@ void KernelBase::setCaliperMgrVariantTuning(VariantID vid,
           { "expr": "any(any#max#Reduction)", "as": "FeatureReduction" },
           { "expr": "any(any#max#Atomic)", "as": "FeatureAtomic" },
           { "expr": "any(any#max#View)", "as": "FeatureView" },
-          { "expr": "any(any#max#MPI)", "as": "FeatureMPI" }
+          { "expr": "any(any#max#MPI)", "as": "FeatureMPI" },
+          { "expr": "any(any#max#Complexity)", "as": "Complexity" }
         ]
```
Comment (michaelmckinsey1) [inline src/common/KernelBase.cpp:664]:
```suggestion
        ],
        "group by": ["Complexity"],
```

**Review by michaelmckinsey1 — CHANGES_REQUESTED** [review]
These changes allow complexity to be recorded as a string with Caliper.

The complexity metric can be read using the `string_attributes=["Complexity"]` argument to `from_caliperreader` for either Hatchet or Thicket.

**Review by michaelmckinsey1 — APPROVED** [review]

**Comment by MrBurmark** [comment]
Thanks @michaelmckinsey1

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**